File: Assets/Scripts/humanAi.js
#pragma strict
	public var navAgent: NavMeshAgent ;
	private var currentTarget:GameObject ;

	public var runSpeed:float ;
	public var speedVariation:float = .2;
	
	
	public var fearMaterial:Material;
	
	
	private var baseSpeed:float ;

	//private var goingtoSafe:boolean  = false;

	public var stayInSafe:boolean ;
	private var isAfraid : boolean = false;
	
	private var currentAnimation;
	
	
function Start () {
	//set Random rotation for visual interest
		transform.rotation = getRandomRotation ();

		//load NavAgent
		navAgent = GetComponent("NavMeshAgent");

		var speedVar : float = navAgent.speed * speedVariation;
		
		//Debug.Log(speedVar);
		
		navAgent.speed = Random.Range(navAgent.speed - speedVar, navAgent.speed + speedVar);
		
		
		//store base navagent speed in variable
		baseSpeed = navAgent.speed;

		var g : GameObject = getRandomNavTarget ("SafeZone");		
		if (g == null) {g = getRandomNavTarget("Finish");}
		
		setTarget(g);
		
		
}

function PlayAnimation(animationName) {
			var a : Animation = GetComponent("Animation");
			a.wrapMode = WrapMode.Loop;
		
			a.Play(animationName.ToString());
			currentAnimation = animationName;

}

function FixedUpdate () {
	if (navAgent.velocity.magnitude >0 && currentAnimation != "walk") {
			PlayAnimation("walk");
			
		}	
	
	//check if target destination has been destroyed
	if (currentTarget == null) {	
			var t : GameObject = getRandomNavTarget ("SafeZone");
			if (t == null) {t = getRandomNavTarget("Finish");}
			
			if (t != null) setTarget(t);
			
			navAgent.speed = baseSpeed;
			//navAgent.SetDestination (currentTarget.transform.position);	
	} else {

		if (navAgent.remainingDistance < 5) {
					if (goingtoSafe() && isAfraid && currentTarget != null) {
						Debug.Log(goingtoSafe);
						enterSafeZone();
					}else {
					
					navAgent.speed = baseSpeed;
					
					var g : GameObject = getRandomNavTarget ("SafeZone");		
					if (g == null) {g = getRandomNavTarget("Finish");}
		
					setTarget(g);
				}
			
			}		
		}
}



function getRandomRotation() {
		var y:int = Random.Range (0, 360);		
		var q:Quaternion = Quaternion.Euler (0, y, 0);		
		return q;
		
	}
	
		
function setTarget (g:GameObject) {
	currentTarget = g;
	navAgent.SetDestination(currentTarget.transform.position);
	
	}	
	
	
function getRandomNavTarget(tagName:String ) {
	var targets: GameObject[]  = GameObject.FindGameObjectsWithTag (tagName);

	if (targets.Length > 0 ) {
		var x:int = Random.Range (0, targets.Length);
	
		return targets [x];
	}
	return null;
}

//Destroy human object
function die() {
		Destroy (gameObject);
		}
		
//Sprint to random safe zone
function Afraid() {
	isAfraid = true;

	//navAgent.Stop ();
	navAgent.speed = runSpeed;
	//navAgent.SetDestination (new Vector3 (0, 0, 0));
	//goingtoSafe = true;
	
	//gameObject.renderer.material = fearMaterial;
	
	var g : GameObject = getRandomNavTarget ("SafeZone");		
	if (g == null) {g = getRandomNavTarget("Finish");}
		
	if (navAgent != null && g != null)  setTarget(g);
		
	
	
}


function goingtoSafe() {
	if (currentTarget.tag == "SafeZone") 
		{return true;}
	else 
		{return false;}
}
	
function enterSafeZone() {
	currentTarget.SendMessage("addHuman");
	
	die();
}

File: Assets/Scripts/humanAi.cs
using UnityEngine;
using System.Collections;

public class humanAI : commonAI, ICanBeScared
{
	// few more publics specific to humans...
	public Material fearMaterial;
	public bool stayInSafe;
	private bool isAfraid = false;

	// Use this for initialization
	public override void Start () 
	{
		// do baseline start actions first...
		base.Start();

		// set Random rotation for visual interest
		// use global static object to utilize smaller memory
		// footprint vs per instance of class
		transform.rotation = gs.getRandomRotation();

		// human navigation targets... look for safe-zones first, then finish zones
		defaultNavTargets.Clear();
		defaultNavTargets.Add(eNavTargets.SafeZone);
		defaultNavTargets.Add(eNavTargets.Finish);

		// get initial target for human...
		moveToNewTarget();
	}

	// Update is called once per frame
	void Update () 
	{
		// if no target, get one now regardless of requiring 
		// "SafeZone" vs "Finish"
		if( currentTarget == null )
			moveToNewTarget();

		// if player is moving, but not walking and they are not in "afraid" mode,
		// then set their animation to a simple walk mode
		if (	navAgent.velocity.magnitude > 0 
			&& ! CurrentAnimation().Equals ( "walk" ) 
			&& ! isAfraid ) 
		{
			navAgent.speed = baseSpeed;
			PlayAnimation("walk");
		}
	
		// if time passed the interval check, 
		// what do we need to do now?
		if( reachedTarget())
		{
			// if the human is afraid and already running to a safe-zone
			// destination, enter them into said safe zone since they are
			// within the range of the stop distance vs remaining distance
			if (isAfraid && currentTarget.tag == eNavTargets.SafeZone.ToString())
			{
				// Tell the target to add a new human represented by THIS
				// one just about to enter, then KILL this instance from
				// the visual screen
				currentTarget.SendMessage("addHuman");
				die();
			}
			else
				moveToNewTarget();
		}
		else 
			// in addition, check for being stagnant (clarification in commonAI.cs)
			IsMovementStagnant();

	}

	
	// Sprint to "safe" location (0,0,0 is test)
	public void Afraid() 
	{
		// if being destroyed, get out, dont do anything else
		if (isDestroying)
			return;

		// set flag that human is afraid, and then set the halo effect
		// which is a coloring under them when running to give visual effect
		// of movement state...
		isAfraid = true;
		GetComponent("Halo").active = true;

		moveToNewTarget();

		// RUN when afraid after new target destination established
		// the HumanHero with child object LittleHero_Solo has
		// animations of walk, idle, sprint
		navAgent.speed = runSpeed;
		PlayAnimation("sprint");
	}
}
